### PR TITLE
Generate the binding redirects for the compiler automatically

### LINKED ISSUE #7612: Auto-generate binding redirects for csc, vbc, csccore, vbccore, compiler server
Currently we list the redirects manually in each .exe.config file.
We should have a single place that lists the redirects and generate .config and AssemblyRedirects.cs files (using msbuild or a custom script).

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This removes the hard-coded binding redirects that exist within our
source tree in favor of redirects that are computed automatically. This
takes advantage of the built-in binding redirect generation support
in MSBuild. In order to get the redirects to merge properly with
the static stuff we have, I renamed the .exe.config files to App.config
which is the name the tooling looks for.

Fixes #7612.

*Review:* @dotnet/roslyn-compiler, @dotnet/roslyn-infrastructure

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-compiler: I see there's another copy of these in the build/bootstrap folder. It seems the only difference is there's an additional trace listener being added. Is that still necessary? Why can't that be configured for everything?

File: build/Targets/VSL.Imports.targets [hunk 1]
```diff
@@ -132,6 +132,26 @@
         <MaxVersion>$(VisualStudioReferenceAssemblyVersion)</MaxVersion>
       </SuggestedBindingRedirects>
   </ItemGroup>
+
+  <ItemGroup Condition="'$(GenerateCompilerBindingRedirects)' == 'true'">
+    <!-- The version 1.0.0.0 in these references doesn't actually matter; the generated versions will redirect from 0.0.0.0 automatically -->
```
Comment (tmat) [inline build/Targets/VSL.Imports.targets]: Can the version be 0.0.0.0 then? That's more obviously a placeholder value.

File: build/Targets/VSL.Imports.targets [hunk 2]
```diff
@@ -132,6 +132,26 @@
         <MaxVersion>$(VisualStudioReferenceAssemblyVersion)</MaxVersion>
       </SuggestedBindingRedirects>
   </ItemGroup>
+
+  <ItemGroup Condition="'$(GenerateCompilerBindingRedirects)' == 'true'">
```
Comment (tmat) [inline build/Targets/VSL.Imports.targets]: Can we move this to a sepate .targets file imported by project that need it? VSL.Imports is becoming a kitchen sink of stuff.

Comment (jasonmalinowski) [inline build/Targets/VSL.Imports.targets]: Good idea, I've updated the PR.

**Comment by jasonmalinowski** [comment]
Ping @dotnet/roslyn-compiler, can somebody clarify what's the best course of action for the build/bootstrap folder?

**Comment by jaredpar** [comment]
test mac please

**Comment by agocke** [comment]
LGTM. I don't see a reason for the bootstrap exe configs. Perhaps we used to be building the bootstrap compiler with version 42.42.42.42 and the purpose of those files was to override the built-in redirects to go to 42.42.42.42?

**Comment by jasonmalinowski** [comment]
@agocke, but the bootstrap compiler configs don't have 42.42.42.42 encoded, so it's not that. The only seeming difference I can see is that they set a different trace listener, presumably to prevent debug asserts from popping up. I guess digging further @jaredpar did this in b6c2e19effe4e8311b26638c5f3caddd7627b171. So (1) is there any reason we don't just want this always (including shipping bits) and (2) if the answer to (1) is no, what's a better approach here?

**Comment by jasonmalinowski** [comment]
Speaking with @jaredpar, our plan will be to remove those by having a #if in the compiler to have the same logic.

**Comment by jasonmalinowski** [comment]
I'll be merging this and addressing that in a followup, since @jcouv also observed an issue today: our compiler extension, if you're building locally, ends up with the shipping redirects, which means analyzers won't work properly because the redirect into nowhere land. This pointed out that I should have used `AssemblyVersion` instead of `RoslynSemanticVersion` in the redirects. I've updated the PR and will merge when clean.

@dotnet-bot, test mac please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot, retest windows_eta_open_prtest please. #11856 struck again.

**Pull request merged**